site/en/docs/extensions/mv3/getstarted/tut-tabs-manager/index.md

```diff
@@ -0,0 +1,382 @@
+  ...
+  "host_permissions": [
+    "https://developer.chrome.com/*"
+  ],
+  ...
+}
+```
+
+#### Query the tabs {: #query }
+
+You can retrieve the tabs from specific URLs using the `tabs.query()` method. Create a `popup.js`
+file and add the following code:
+
+{% Label %}popup.js:{% endLabel %}
+
+```js
+const tabs = await chrome.tabs.query({
+  url: [
+    "https://developer.chrome.com/docs/webstore/*",
+    "https://developer.chrome.com/docs/extensions/*",
+  ],
+});
+...
+```
+
+{% Details %}
+{% DetailsSummary %}
+💡 **Can I use Chrome APIs directly in the popup?**
+{% endDetailsSummary %}
+
+A popup and other extension pages can call any [Chrome API][doc-apis] because they are served from the
+chrome schema. For example `chrome-extension://EXTENSION_ID/popup.html`.
+
+{% endDetails %}
+
+#### Focus on a tab {: #focus}
+
+First, the extension will sort the list alphabetically. Then, when a list item is clicked, it will
+focus on that tab using `tabs.update()` and bring the window to the front using `windows.update()`.
+Add the following code to `popup.js`:
+
+{% Label %}popup.js:{% endLabel %}
+
+```js
+...
+const collator = new Intl.Collator();
+tabs.sort((a, b) => collator.compare(a.title, b.title));
+
+const template = document.getElementById("li_template");
+const elements = new Set();
+for (const tab of tabs) {
+  const element = template.content.firstElementChild.cloneNode(true);
+
+  const title = tab.title.split("-")[0].trim();
+  const pathname = new URL(tab.url).pathname.slice("/docs".length);
+
+  element.querySelector(".title").textContent = title;
+  element.querySelector(".pathname").textContent = pathname;
+  element.querySelector("a").addEventListener("click", async () => {
+    // need to focus window as well as the active tab
+    await chrome.tabs.update(tab.id, { active: true });
+    await chrome.windows.update(tab.windowId, { focused: true });
+  });
+
+  elements.add(element);
+}
+document.querySelector("ul").append(...elements);
+...
+```
+
+{% Details %}
+{% DetailsSummary %}
+💡 **Interesting JavaScript used in this code**
+{% endDetailsSummary %}
+
+- The [Collator](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Intl/Collator) is used to sort the tabs array by title in any language.
+- The [template tag](https://web.dev/webcomponents-template/) is used to
+  define an HTML element that can be cloned instead of using `document.createElement()` to create each item.
+- The [URL constructor](https://developer.mozilla.org/docs/Web/API/URL/URL) is used to create and parse URLs.
+
+{% endDetails %}
+
+#### Group the tabs {: #group}
+
+Add the following code to create a button that will group all the tabs using [`tabs.group()`]() and
+move them into the current window.
+
+{% Label %}popup.js:{% endLabel %}
+
+```js
+...
+const button = document.querySelector("button");
+button.addEventListener("click", async () => {
+  const group = await chrome.tabs.group({ tabIds: tabs.map(({ id }) => id) });
```

Comment:
For clarity, I think it would be better to split the `chrome.tabs.group()` call and the `tabs.map()` call. This approach allows the developer to place a breakpoint on the `const group` line and see the value of `tabIds`.

```suggestion
  const tabIds = tabs.map(({ id }) => id);
  const group = await chrome.tabs.group({ tabIds });
```

Reply:
Alternatively, this makes it more clear what the entire object will be at debug time.

```suggestion
  const groupOptions = {
    tabIds: tabs.map(({ id }) => id) }),
  };
  const group = await chrome.tabs.group(groupOptions);
```

Between the two approaches, I have a slight personal preference for my first suggestion.

Reply:
Agreed, let's go with the first suggestion.